### PR TITLE
Fixes SplineSmoothing bugs and executes the algorithm faster

### LINKED ISSUE #13335: SplineSmoothing takes forever for some curves that are not terribly difficult
Take the first spectrum of the file `ENGINX_precalculated_vanadium_run000236516_bank_curves.nxs` which can be found in `Data/DocTest`

Apply the algorithm [`SplineSmoothing`](http://docs.mantidproject.org/nightly/algorithms/SplineSmoothing-v1.html). I run it as a test to see how it works, but I had to stop it after over two hours of calculations using 2/3 cores.

Investigate why this happens. In principle the algorithm works very well for the example shown in the doc test, even though the example is small. Check how well the algorithm scales when the number of points / bins increases.

### PR DESCRIPTION
**Pull request by shahrozahmed1**
Fixes #13335

The algorithm now allows user to include number of breaking points which will enable the algorithm to execute functions with large number of spikes or noise. The lower breaking points number set the faster algorithm will execute the function but obviously wont produce high quality smoothing graph.

_Also updated the SplineSmoothing-v1 documentation_

**To Test:**
Take the first spectrum of the file `ENGINX_precalculated_vanadium_run000236516_bank_curves.nxs` which can be found in `Data/DocTest`

Apply the algorithm [`SplineSmoothing`](http://docs.mantidproject.org/nightly/algorithms/SplineSmoothing-v1.html).
set `MaxNumberOfBreaks` to: **20**

### CODE REVIEW

File: Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp [hunk 1]
```diff
@@ -302,29 +309,41 @@ void SplineSmoothing::addSmoothingPoints(const std::set<int> &points,
...
+  }
   // number of points to start with
-  int numSmoothPts(M_START_SMOOTH_POINTS);
+  int numSmoothPts(maxBreaks);
 
   // evenly space initial points over data set
   int delta = xSize / numSmoothPts;
-  for (int i = 0; i < xSize; i += delta) {
+  
+  g_log.information() << "delta is: " << delta << std::endl;
```
Comment (FedeMPouzols) [inline Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp]: Remember to remove this log message for the final version.

If in this or any algorithm you want to leave some detailed logs that are important to track how the algorithm runs you could put those as "debug" level logs.

File: Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp [hunk 2]
```diff
@@ -302,29 +309,41 @@ void SplineSmoothing::addSmoothingPoints(const std::set<int> &points,
...
+  
+  for (int i = 0; i < xSize; i+=delta) {
     smoothPts.insert(i);
   }
   smoothPts.insert(xSize - 1);
+  
+  g_log.information() << "smoothPtrs has: " << smoothPts.size() << " elements"
+                      << std::endl;
```
Comment (FedeMPouzols) [inline Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp]: Same here.

File: Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp [hunk 3]
```diff
@@ -302,29 +309,41 @@ void SplineSmoothing::addSmoothingPoints(const std::set<int> &points,
...
 
   bool resmooth(true);
   while (resmooth) {
     // if we're using all points then we can't do anything more.
-    if (smoothPts.size() >= xs.size() - 1)
+    if (smoothPts.size() > maxBreaks + 2)
       break;
 
+    g_log.information() << "iteration inside resmooth " << std::endl;
+
```
Comment (FedeMPouzols) [inline Code/Mantid/Framework/CurveFitting/src/SplineSmoothing.cpp]: Same here.

**Comment by wdzhou** [comment]
It works well with the new feature.  Thus it is closed.

**Pull request merged**